### DISCUSSION TITLE
Guidance regarding YALMIP solution for the following problem

### DISCUSSION THREAD

**Discussion by siriprasad1331** [yalmip/YALMIP#983] [Optimization modelling]
Hello Sir, 

My name is Prasad, I am writing this mail seeking your help in solving the LMI problem, using YALMIP Toolbox.

The problem is to derive the controller parameters solving LMI formulation. But the LMI formulations has been framed based on bounded real lemma and multi objective design problem. I am unable to formulate the same using YALMIP toolbox. I am able to derive the controller parameters using H infinity synthesis, but bounded real lemma formulation, I am unable to get. 

I am placing the snapshot of the equations
The system is given by
![image](https://user-images.githubusercontent.com/90329614/132519337-bc1c8b79-618d-470f-9072-a60142c920e3.png)
CLosed loop system becomes

![image](https://user-images.githubusercontent.com/90329614/132519445-9d4129f6-b4d9-4aaf-8bb4-465926447446.png)

Transfer function is defined as

![image](https://user-images.githubusercontent.com/90329614/132519512-ce7f3762-edcb-424f-8a6e-8b174546889f.png)

Optimization problem becomes
![image](https://user-images.githubusercontent.com/90329614/132519615-c45ed04c-b40f-464d-b299-e30e02fc0694.png)

Bounded real lemma characterization becomes 

![image](https://user-images.githubusercontent.com/90329614/132520071-194609ec-57c3-4b6c-8cdb-fddba5a6ddd7.png)

![image](https://user-images.githubusercontent.com/90329614/132520310-f2b70f0e-c292-4023-b47c-2e23a52b6da7.png)

Finally LMI formualtion

![image](https://user-images.githubusercontent.com/90329614/132520561-8e2594bc-f487-4258-a357-465c0b7a9164.png)

![image](https://user-images.githubusercontent.com/90329614/132520612-505d2786-e4af-49d8-a040-9b60b03c8f44.png)
 

Looking forward for your guidnace

**Comment by johanlofberg** [comment]
So what is the problem? (20) which I assume you're solving looks standard
https://yalmip.github.io/tutorial/semidefiniteprogramming/

**Reply by siriprasad1331** [reply]
Sir, the problem is
For the realization of equation 20, I need matrices A^ B^, C^ and D^. Which are defined based upon equation 14 and 15. In equation 14 ann15 we require the controller matrices Ak, Bk, Ck, and Dk. 
Now when I am doing H infinity synthesis for equation 16, I am getting gamma of 5.6. and from that I can get the matrices Ak, Bk, Ck, and Dk.
Now my doubt is shall I use those, to define first A^ B^ .. and then put into LMI represented in 20? Is that correct method? Also, when I am doing, H infinity for 16, I am getting large equation, how to separate 14 and 15 out of that?
If I get Ak.. there then what is the relevance of equation 21?
and how to define Y, X, M and N matrix?

**Reply by johanlofberg** [reply]
Your decision variables are X,Y,A^,B^ tc so you define those, define the big matrices in 20 and add those constraints, solve, extract values of optimal solutions, and then compute Ak etc from those

**Reply by siriprasad1331** [reply]
Sir, but as per equation 19, to define A^, in the matrix shown by (20), we need Ak, Bk,,, etc. So shall we first derive Ak, Bk.. to define A^.. and then put in (20)?
If that is the case what does (21) defines sir?
As per equation we also need N, M, Y, C to define A^, B^.. etc. how to get those sir?
sorry if my understanding is wrong.

**Reply by johanlofberg** [reply]
19 is the definition of a variable change

**Reply by siriprasad1331** [reply]
Ok Sir, I will try to code variable change and try to put in 20. Thanks for your support.

**Reply by johanlofberg** [reply]
You don't code any variable change. You declare the decision variables I said, define 20, solve, extract solution and then use 21 to compute original variables

Hard problem (bilinear in x and k) a*x + b*k*x <0. Variable change z = k*x. New simple problem (in x and z) a*x + b*z < 0. Solve leads e.g.  to x= -10, z = 65. recover k = 65/-10. Exactly the same procedure

**Reply by siriprasad1331** [reply]
Ok Sir, I will do and get back to you.

**Reply by siriprasad1331** [reply]
Sir, can you provide any reference for hard problem(bilinear in x and k)?

**Reply by johanlofberg** [reply]
The very problem you are solving starts out as a nasty BMI, but then some clever person (Scherer I guess) figures out how to perform a variable change to render it linear, solve, and recover original variables

**Reply by siriprasad1331** [reply]
yes.. Carsten Scherer.

**Reply by siriprasad1331** [reply]
Sir, hope this mail find you good. In continuation to my previous query, I am getting error after in declaring sdpvar/horzcat as per error message. The code I have used is 
wo = 314.15;
Ct = 100e-6;
Lt = 100e-6;
Rt = 1.5e-3;
A = [0 wo 1/Ct 0;
    -wo 0 0 1/Ct;
    -1/Lt 0 -(Rt/Lt) wo;
    0 -1/Lt wo -(Rt/Lt)];
B1 = [0 0;
    0 0;
    1/Lt 0;
    0 1/Lt];
B2 = [-1/Ct 0;
       0 -1/Ct;
       0 0;
       0 0];
C = [1 0 0 0;
     0 1 0 0];
D = [0 0;
     0 0];
H1 = ss(A,B1,C,D);
G1 = tf(H1);
H2 = ss(A,B2,C,D);
G2 = tf(H2);
num11 = [0.33 50];
den11 = [1 0.05];
W11 = tf(num11,den11);
W1 = [W11 0;
      0 W11];
num22 = [1 100];
den22 = [1 10000];
W22 = tf(num22,den22);
W2 = [W22 0 0 0;
      0 W22 0 0];
num33 = [56850];
den33 = [1 30.16 56850];
WD11 = tf(num33,den33);
WD = [WD11 0;
     0 WD11];
 G = [G1 G2];
P=augw(G,W1,W2,WD);
sys =ss(P);
[a,b,c,d] = ssdata(sys)
A = a;
BW = mat2cell(b,[16],[2 2 2]);
BW1 = BW{1,1};
BW2 = BW{1,2};
B = BW{1,3};
CZ = mat2cell(c,[3 3 2],[16]);
CZ1 = CZ{1,1};
CZ2 = CZ{2,1};
C = CZ{3,1};
DZ = mat2cell(d,[3 3 2],[2 2 2]);
DWZ1 = DZ{1,1};
DWZ12 = DZ{1,2};
DZ1 = DZ{1,3};
DWZ21 = DZ{2,1};
DWZ2 = DZ{2,2};
DZ2 = DZ{2,3};
DW1 = DZ{3,1};
DW2 = DZ{3,2};
DZZ = DZ{3,3};
clear('yalmip')
X = sdpvar(16);
Y = sdpvar(16);
Ah = sdpvar(16);
Bh = sdpvar(16,2,'full');
Ch = sdpvar(2,16,'full');
Dh = sdpvar(2,2,'full');
sdpvar gam
 
LMI1=blkvar;
LMI1(1,1)=A*X+X*A'+B*Ch+Ch'*B';
LMI1(1,2)=Ah'+A+B*Dh*C;
LMI1(1,3)=BW1'+DW1'*Dh'*B';
LMI1(1,4)=X*CZ1'+Ch'*DZ1';
LMI1(2,2)=Y*A+A'*Y+Bh*C+C'*Bh';
LMI1(2,3)=Y*BW1+B*DW1;
LMI1(2,4)=CZ1'+C'*Dh'*DZ1';
LMI1(3,3)=-gam*eye(4);
LMI1(3,4)=DWZ1';
LMI1(4,4)=-gam*eye(4);
LMI1=sdpvar(LMI1);

LMI2=blkvar;
LMI2(1,1)=A*X+X*A'+B*Ch+Ch'*B';
LMI2(1,2)=Ah'+A+B*Dh*C;
LMI2(1,3)=BW2'+DW2'*Dh'*B';
LMI2(1,4)=X*CZ2'+Ch'*DZ2';
LMI2(2,2)=Y*A+A'*Y+Bh*C+C'*Bh';
LMI2(2,3)=Y*BW2+B*DW2;
LMI2(2,4)=CZ2'+C'*Dh'*DZ2';
LMI2(3,3)=-eye(4);
LMI2(3,4)=DWZ2';
LMI2(4,4)=-eye(4);
LMI2=sdpvar(LMI2);

Con=[LMI1<=0,LMI2<=0, X>0,Y>0,gam>=0.1];
solvesdp(Con) 
gam = double(gam)
X = double(X)
Y = double(P)
Ah = double(P)
Bh = double(P)
Ch = double(P)
Dh = double(P)


The error I am getting is
Error using sdpvar/horzcat (line 36)
All matrices on a row in the bracketed expression must have the same number of rows.

Error in blkvar/sdpvar (line 62)
            Y = [Y X.blocks{i,j}];

Error in invlmi1 (line 84)
LMI1=sdpvar(LMI1);

As you have asked last time, I have declared the variable, Ah,Bh,Ch, Dh, X and Y. The dimensions of the variables I have taken based on augmented matrix P. The error showing, I am unable to debug. 

Sir, could you please suggest me, where I am going wrong?

Secondly, in the expression (19), how to get the values of M and N, since I have not defined them as variables. They are not  in the expression (20). They are defined as

![5](https://user-images.githubusercontent.com/90329614/140594176-fcb39d32-766a-459b-b89a-d4176fa8cf73.PNG)

Thank you..

**Reply by johanlofberg** [reply]
You have to put the code in code blocks when pasting it here, it does not run when copied to MATLAB otherwise

But the error message is pretty clearly stated. The dimensions of the blocks do not match, you're trying to do something similar to

```
>> clear all
>> B = blkvar;
>> B(1,1)=sdpvar(2);
>> B(1,2)=zeros(3,2);
>> sdpvar(B)
Error using sdpvar/horzcat (line 36)
All matrices on a row in the bracketed
expression must have the same number of
rows.
```

**Comment by siriprasad1331** [comment]
Ok sir, I will get you back.

Regards
